app/models/client_application.rb

```diff
@@ -39,6 +39,7 @@ class ClientApplication < ApplicationRecord
 
   validates :key, :presence => true, :uniqueness => true
   validates :name, :url, :secret, :presence => true
+  # TODO: Consider using UrlValidator from validate_url gem.
```

Comment:
Turns out I've considered this before! But I'd forgotten about that. Unfortunately it won't work. `validate_url` requires a list of permitted schemes, but for oauth callbacks any scheme needs to be allowed. See https://github.com/perfectline/validates_url/issues/99

I think this comment can be removed.

Reply:
removed